generic-ec/src/multiscalar.rs

```diff
@@ -0,0 +1,406 @@
+/// can be rewritten as:
+///
+/// $$
+/// \begin{aligned}
+/// s_1 P_1 &=&& s_{1,0} P_1 &&+&& 16^1 s_{1,1} P_1 &&+ \dots +&& 16^{k-1} s_{1,k-1} P_1 \\\\
+///    \+   & &&        +    && &&             +    &&         &&                   +    \\\\
+/// s_2 P_2 &=&& s_{2,0} P_2 &&+&& 16^1 s_{2,1} P_2 &&+ \dots +&& 16^{k-1} s_{2,k-1} P_2 \\\\
+///    \+   & &&        +    && &&             +    &&         &&                   +    \\\\
+/// \vdots  & && \vdots      && && \vdots           &&         && \vdots                 \\\\
+///    \+   & &&        +    && &&             +    &&         &&                   +    \\\\
+/// s_n P_n &=&& s_{n,0} P_n &&+&& 16^1 s_{n,1} P_n &&+ \dots +&& 16^{k-1} s_{n,k-1} P_n
+/// \end{aligned}
+/// $$
+///
+/// Straus algorithm computes the sum column by column from right to left, multiplying result
+/// by 16 after each column sum is computed. Also, it uses the precomputed table
+/// $T_i = [\O, P_i, 2 P_i, \dots, 15 P_i]$ to optimize multiplication $s_{i,j} P_i$.
+///
+/// Transformed sum that's computed by Straus algorithm looks like this:
+///
+/// $$
+/// \begin{aligned}
+/// Q_{k-1} &=             & T_{1,s_{1,k-1}} &+ \dots &+ T_{n,s_{n,k-1}}& \\\\
+/// Q_{k-2} &= 16 Q_{k-1} +& T_{1,s_{1,k-2}} &+ \dots &+ T_{n,s_{n,k-2}}& \\\\
+/// \vdots  &              &                 &        &                 & \\\\
+/// Q_1     &= 16 Q_2     +& T_{1,s_{1,1}}   &+ \dots &+ T_{n,s_{n,1}}  & \\\\
+/// Q = Q_0 &= 16 Q_1     +& T_{1,s_{1,0}}   &+ \dots &+ T_{n,s_{n,0}}  &
+/// \end{aligned}
+/// $$
+#[cfg(feature = "alloc")]
+pub struct Straus;
+
+#[cfg(feature = "alloc")]
+impl<E: Curve> MultiscalarMul<E> for Straus {
+    fn multiscalar_mul<S, P>(scalar_points: impl IntoIterator<Item = (S, P)>) -> Point<E>
+    where
+        S: AsRef<Scalar<E>>,
+        P: AsRef<Point<E>>,
+    {
+        let (mut scalars, points): (Vec<Radix16Iter<E>>, Vec<Point<E>>) = scalar_points
+            .into_iter()
+            .map(|(scalar, point)| (scalar.as_ref().as_radix16_be(), *point.as_ref()))
+            .unzip();
+        Self::multiscalar_mul_inner(&mut scalars, &points)
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl Straus {
+    fn multiscalar_mul_inner<E: Curve>(
+        scalars: &mut [Radix16Iter<E>],
+        points: &[Point<E>],
+    ) -> Point<E> {
+        if scalars.is_empty() {
+            return Point::zero();
+        }
+
+        // table[i] = [point_i, 2 * point_i, ..., 15 * point_i]
+        let table = points
+            .iter()
+            .map(|point_i| {
+                iter::successors(Some(*point_i), |point| Some(point + point_i))
+                    .take(15)
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
+
+        // `serialized_len` is amount of radix256-digits. We multiply it by 2 and get
+        // amount of radix16-digits
+        let num_digits = 2 * Scalar::<E>::serialized_len();
```

Comment:
Alternative: implement [ExactSizeIterator](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html) for Radix16Iter

Reply:
> Note that this trait is a safe trait and as such does not and cannot guarantee that the returned length is correct

Reply:
Ah I thought this comment was referring to another thing. Yes we can do that. Since it would be our implementation, we can trust that it guarantees that the returned length is correct

Reply:
I implemented [ExactSizeIterator](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html) for Radix16Iter as you suggested, see if it looks fine